x/inflation/types/v1beta2/inflation_calculator.go

```diff
@@ -0,0 +1,83 @@
+package v1beta2
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+	paramstypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/ericlagergren/decimal"
+)
+
+func GetInflationCalculator(
+	genesisTime time.Time,
+	inflationParamSubspace paramstypes.Subspace,
+) minttypes.InflationCalculationFn {
+	return func(ctx sdk.Context, minter minttypes.Minter, params minttypes.Params, bondedRatio sdk.Dec) sdk.Dec {
+		var inflationParams Params
+		inflationParamSubspace.GetParamSet(ctx, &inflationParams)
+
+		// years passed since genesis = seconds passed since genesis / number of seconds per year
+		// can be a fraction, eg: 0.5
+		yearsPassed := decimal.WithPrecision(sdk.Precision).
+			Quo(
+				// seconds since genesis
+				decimal.WithPrecision(sdk.Precision).
+					Sub(
+						decimal.New(ctx.BlockTime().Unix(), 0),
+						decimal.New(genesisTime.Unix(), 0),
+					),
+				// Number of hours in an year = 8766 (averaging the leap year hours)
+				// Number of minutes in an hour = 60
+				// Number of seconds in a minute = 60
+				// => Number of seconds per year = 60 * 60 * 8766 = 31557600
+				decimal.New(31557600, 0),
+			)
+		// 2^(-t/tHalf)
+		pow := decimal.WithPrecision(sdk.Precision)
+		pow = pow.Context.
+			Pow(
+				pow,
+				decimal.New(2, 0),
+				decimal.WithPrecision(sdk.Precision).
+					Mul(
+						decimal.New(-1, 0),
+						decimal.WithPrecision(sdk.Precision).
+							Quo(
+								yearsPassed,
+								decimal.New(int64(inflationParams.InflationDecayFactor), 0),
+							),
+					),
+			)
+		idealInflation := inflationParams.InitialInflation.Mul(sdk.MustNewDecFromStr(pow.String()))
```

Comment:
Use [`Int()`](https://pkg.go.dev/github.com/ericlagergren/decimal?utm_source=godoc#Big.Int) or [`Float()`](https://pkg.go.dev/github.com/ericlagergren/decimal?utm_source=godoc#Big.Float).

Multiply then divide by precision if needed.

Reply:
`inflationParams.InitialInflation` is a `sdk.Dec`. `sdk.Dec` has 18 precision digits.
`pow` is a `decimal.Big`. I am using the same precision of 18 digits for `decimal.Big` everywhere.

So, converting `pow` to string and parsing it into `sdk.Dec` won't lose any precision.

Could you please elaborate why and where should we use `Int()/Float()`? Also, why and where we need to `multiply and divide by precision`?

Reply:
We don't need that much precision, especially for the final result.

If the result is less than one and we use 6 units of precision then you can make an sdk.Decimal by with something like:

`sdk.Decimal(math.Int(val * 10^6)) / 10^6`

Reply:
Done